src/include/processor/operator/persistent/reader.h

```diff
@@ -57,15 +57,11 @@ class Reader : public PhysicalOperator {
     void readNextDataChunk();
 
     template<ReaderSharedState::ReadMode READ_MODE>
-    inline void lockForSerial() {
+    [[nodiscard]] inline auto lockForSerial() {
```

Comment:
We always explicitly write the return type instead of auto in function declaration/implementation.
I am a little bit confused about the return type:
 std::lock_guard and nullptr has different type, how can we return two different types in the same function?
BTW: We never use nodiscard in the code base,  should we use it from now on?

Reply:
`auto` is used to enable this confusing behaviour. We have no lock to hold if `READ_MODE` is parallel, so we return a dummy object.

Since this is a templated function, it can have *two return values depending on READ_MODE*.

Another way to write this would be using template specialization, which might be easier to understand. Thoughts?

On the topic of `[[nodiscard]]`, it's a particularly nice convention here because it makes it obvious that `lockForSerial()` shouldn't be ignored (since it's a lock guard).

Should we use it everywhere? Maybe. Continuing to add things to bulletproof our codebase is a good idea, and this is one of them. There's even a `clang-tidy` [check](https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-nodiscard.html) for using `[[nodiscard]]` on all `const` functions (since they have no way of modifying the object, and discarding the result is likely an error).

Reply:
I would prefere using template specialization, this function is too confusing.

Reply:
Seems template specialization won't work since the signatures have to match. I'll use a `unique_ptr`.

Reply:
Updated to use `std::optional`